js/style/style.js

```diff
@@ -638,7 +638,16 @@ Style.prototype = util.inherit(Evented, {
         }
 
         var sourceResults = [];
-        for (var id in this.sources) {
+        var sourcesToQuery = {};
+        if (params.layers) {
+            for (var i = 0; i < params.layers.length; i++) {
+                var sourceId = this._layers[params.layers[i]].source;
+                sourcesToQuery[sourceId] = true;
+            }
+        } else {
+            sourcesToQuery = this.sources;
+        }
+        for (var id in sourcesToQuery) {
```

Comment:
Sure, 👍 for filtering the sources within the loop like you propose.

I'm a little bit wary of using `array.some()` here rather than a plain for loop over `params.layers`, only because this method could get called with a pretty high frequency, e.g., in a `mousemove`. handler, where this might make a difference.  (But I admit this is based on intuition, not quantitative evidence)